### PR TITLE
Update Arrow 45.0.0 And Datum Arithmetic, change Decimal Division semantics

### LINKED ISSUE #6828: Incorrect Decimal Division Coercion
### Describe the bug

Currently when performing decimal division `coercion_decimal_mathematics_type` as called by `BinaryExpr::evaluate` will coerce both inputs to the wider precision type. The division kernel will then scale the left hand side by the output scale. 

To see why this is an issue consider

```
1 DECIMAL(38, 20) / 5 DECIMAL(38, 0)
```

This computation shouldn't overflow as it can just perform `10e20 / 5`

The issue is that the coercion logic will scale the right hand side to `5e20` requiring it to also scale the left hand side to `10e40` which will overflow

### To Reproduce

```
❯ create table foo (a DECIMAL(38, 20), b DECIMAL(38, 0));
0 rows in set. Query took 0.001 seconds.
❯ insert into foo VALUES (1, 1000000000000);
+-------+
| count |
+-------+
| 1     |
+-------+
1 row in set. Query took 0.003 seconds.
❯ select a / b from foo;
Decimal128(38, 20) vs Decimal128(38, 20)
+------------------------------------------+
| foo.a / foo.b                            |
+------------------------------------------+
| 0.00000000000000000000000000000000743222 |
+------------------------------------------+
1 row in set. Query took 0.004 seconds.
```

### Expected behavior

_No response_

### Additional context

https://github.com/apache/arrow-datafusion/issues/6794

### LINKED ISSUE #6933: Incorrect Interval Negation
### Describe the bug

Negating an array of intervals yields an incorrect result. This is caused by https://github.com/apache/arrow-rs/issues/4488

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #6828
Closes #6933
Closes https://github.com/apache/arrow-datafusion/issues/7068

# Rationale for this change

Use new kernels added in https://github.com/apache/arrow-rs/pull/4465

Make timestamp arithmetic kernels more precse

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

* Timestamp arithmetic will now return durations
* Adding dates to negative intervals no longer works (support was only added in #6778 largely by accident)

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/interval.slt [hunk 1]
```diff
@@ -430,15 +430,11 @@ select '1 month'::interval + '1980-01-01T12:00:00'::timestamp;
 ----
 1980-02-01T12:00:00
 
-query D
+query error DataFusion error: Error during planning: Cannot coerce arithmetic expression Interval\(MonthDayNano\) \- Date32 to valid types
```
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/interval.slt:433]: This query was only supported following #6778 and was done so accidentally - https://github.com/apache/arrow-datafusion/pull/6778#discussion_r1243994954

I tested some other systems, duckdb explicitly doesn't support this, and I couldn't work out how to get mysql to express this, so I think this is fine

File: datafusion/core/tests/sql/timestamp.rs [hunk 2]
```diff
@@ -1035,13 +1035,13 @@ async fn timestamp_sub_with_tz() -> Result<()> {
...
-        "| val | ts_diff                                           |",
-        "+-----+---------------------------------------------------+",
-        "| 3   | 0 years 0 mons 0 days 10 hours 0 mins 30.000 secs |",
-        "| 1   | 0 years 0 mons 0 days 10 hours 0 mins 20.000 secs |",
-        "| 2   | 0 years 0 mons 0 days 10 hours 0 mins 10.000 secs |",
-        "+-----+---------------------------------------------------+",
+        "+-----+---------+",
+        "| val | ts_diff |",
+        "+-----+---------+",
+        "| 3   | PT30S   |",
```
Comment (tustvold) [inline datafusion/core/tests/sql/timestamp.rs:1041]: Timestamp arithmetic now returns the appropriate duration type

File: datafusion/common/src/scalar.rs [hunk 3]
```diff
@@ -5912,285 +5859,6 @@ mod tests {
         check_array(array);
     }
 
-    fn get_timestamp_test_data(
```
Comment (tustvold) [inline datafusion/common/src/scalar.rs:5915]: This test is removed as timestamp arithmetic no longer returns intervals (which avoids this whole mess of edge cases)

File: datafusion/core/tests/sqllogictests/test_files/timestamps.slt [hunk 4]
```diff
@@ -1182,10 +1182,10 @@ create table bar (val int, i1 interval, i2 interval) as values
 query I?
 SELECT val, ts1 - ts2 FROM foo ORDER BY ts2 - ts1;
 ----
-4 0 years 0 mons -15250 days -13 hours -28 mins -44.999876545 secs
-3 0 years 0 mons 15952 days 23 hours 22 mins 12.667123455 secs
-2 0 years 0 mons 8406 days 1 hours 1 mins 54.877123455 secs
-1 0 years 0 mons 53 days 16 hours 0 mins 20.000000024 secs
+3 P15952DT84132.667123455S
```
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/timestamps.slt:1185]: This does highlight another fun quirk of intervals - they sort strangely

**Comment by alamb** [comment]
🤯 
Screenshot 2023-07-08 at 1 22 32 PM

File: datafusion/physical-expr/src/expressions/negative.rs [hunk 5]
```diff
@@ -236,9 +179,6 @@ mod tests {
         test_array_negative_op!(Int64, 23456i64, 12345i64);
         test_array_negative_op!(Float32, 2345.0f32, 1234.0f32);
         test_array_negative_op!(Float64, 23456.0f64, 12345.0f64);
-        test_array_negative_op_intervals!(YearMonth, 2345i32, 1234i32);
-        test_array_negative_op_intervals!(DayTime, 23456i64, 12345i64);
-        test_array_negative_op_intervals!(MonthDayNano, 234567i128, 123456i128);
```
Comment (tustvold) [inline datafusion/physical-expr/src/expressions/negative.rs:241]: This is moved into sqllogic, unfortunately I couldn't find a way to get a MonthDayNano type, but this is better than nothing

Comment (alamb) [reply datafusion/physical-expr/src/expressions/negative.rs:241]: I think you can do something like this

```
select arrow_cast(interval '30 minutes', 'Interval(MonthDayNano)');
```

Comment (tustvold) [reply datafusion/physical-expr/src/expressions/negative.rs:241]: Arrow cast doesn't appear to support intervals in my experiments

File: datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt [hunk 6]
```diff
@@ -303,9 +303,11 @@ select arrow_cast('30 minutes', 'Interval(MonthDayNano)');
 
 ## Duration
 
-query error DataFusion error: Optimizer rule 'simplify_expressions' failed\ncaused by\nThis feature is not implemented: Can't create a scalar from array of type "Duration\(Second\)"
+query ?
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/arrow_typeof.slt:306]: 🎉

File: datafusion/core/tests/sqllogictests/test_files/binary.slt [hunk 7]
```diff
@@ -110,14 +110,19 @@ ff0102 FixedSizeBinary(3)
 000102 FixedSizeBinary(3)
 
 # Comparison
-# https://github.com/apache/arrow-rs/pull/4492
-query error DataFusion error: Arrow error: Compute error: eq_dyn_binary_scalar only supports Binary or LargeBinary arrays
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/binary.slt:114]: FYI @maxburke

File: datafusion/core/tests/sqllogictests/test_files/decimal.slt [hunk 8]
```diff
@@ -365,7 +365,7 @@ select c1*c5 from decimal_simple;
 query T
 select arrow_typeof(c1/cast(0.00001 as decimal(5,5))) from decimal_simple limit 1;
 ----
-Decimal128(21, 12)
+Decimal128(19, 10)
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/decimal.slt:368]: cc @liukun4515 and @viirya -- I assume you have reviewed the decimal changes upstream

File: datafusion/core/tests/sqllogictests/test_files/interval.slt [hunk 9]
```diff
@@ -430,15 +436,11 @@ select '1 month'::interval + '1980-01-01T12:00:00'::timestamp;
...
 select '1 month'::interval - '1980-01-01'::date;
-----
-1979-12-01
 
-query P
+query error DataFusion error: Error during planning: Cannot coerce arithmetic expression Interval\(MonthDayNano\) \- Timestamp\(Nanosecond, None\) to valid types
 select '1 month'::interval - '1980-01-01T12:00:00'::timestamp;
-----
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/interval.slt:440]: I agree an interval minus a timestamp doesn't make a lot of sense

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 10]
```diff
@@ -2609,14 +2441,14 @@ mod tests {
             Operator::Divide,
             create_decimal_array(
                 &[
-                    Some(99193548387), // 0.99193548387
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:2612]: why these changes?

File: datafusion/physical-expr/src/expressions/datetime.rs
Comment (alamb) [inline datafusion/physical-expr/src/expressions/datetime.rs:1]: 👏

**Review by alamb — COMMENTED** [review]
Thank you so much @tustvold  -- 

giphy


I think this PR looks (more than) great except for the default formatting of durations. Is there any way we can fix that so they aren't in the  somewhat inscrutable form of `P53DT57620.000000024S`?


cc @berkaysynnada and @avantgardnerio  as I think you were involved in the original timestamp code implementation

File: datafusion/core/tests/sql/timestamp.rs (hunk 2 above)
Comment (tustvold) [reply datafusion/core/tests/sql/timestamp.rs:1041]: I would be happy to change the formatting upstream if there is consensus on what it should be. We are currently just doing what chrono does, which appears to be [ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Time_intervals)

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 10 above)
Comment (tustvold) [reply datafusion/physical-expr/src/expressions/binary.rs:2612]: Because the rules for precision of decimal division are now correct #6828

File: datafusion/core/tests/sql/timestamp.rs (hunk 2 above)
Comment (waitingkuo) [reply datafusion/core/tests/sql/timestamp.rs:1041]: i think the coercion rules might not work well. 
```bash
❯ select ('2023-01-02'::timestamp - '2023-01-01'::timestamp)::interval;
+--------------------------------------------------------+
| Utf8("2023-01-02") - Utf8("2023-01-01")                |
+--------------------------------------------------------+
| 0 years 0 mons 0 days 24 hours 0 mins 0.000000000 secs |
+--------------------------------------------------------+
1 row in set. Query took 0.002 seconds.
```
it's converted to (0 month, 0 day, 86400000000000 nanos) instead of (0 month, 1 day, 0 nanos)

which isn't consist with 
```bash
❯ select ('2023-01-02'::timestamp - '2023-01-01'::timestamp);
+-----------------------------------------+
| Utf8("2023-01-02") - Utf8("2023-01-01") |
+-----------------------------------------+
| P1D                                     |
+-----------------------------------------+
1 row in set. Query took 0.002 seconds.
``` 

both formats are fine so we. in case we decide to align with interval format, we should discuss whether it outputs 
`0 years 0 mons 1 days 0 hours 0 mins 0 secs` or 
`0 years 0 mons 0 days 24 hours 0 mins 0 secs` in this case

Comment (tustvold) [reply datafusion/core/tests/sql/timestamp.rs:1041]: I think we should probably change duration output to be a number of the duration unit, i.e. seconds, milliseconds, etc... to avoid potential ambiguity over what the duration represents. It might be confusing if both intervals and durations report some number of days, but arithmetic involving them has different semantics across DST boundaries

**Comment by waitingkuo** [comment]
@berkaysynnada @tustvold @alamb 

currently we couldn't covert interval back to duration if it contains mons or days

```bash
❯ select arrow_cast(interval '1 day', 'Duration(Nanosecond)');
Optimizer rule 'simplify_expressions' failed
caused by
Arrow error: Compute error: Cannot cast to Duration(Nanosecond). Overflowing on 18446744073709551616
```

however pure nanos representation works

```bash
❯ select arrow_cast(interval '24 hours', 'Duration(Nanosecond)');
+----------------------------------------+
| IntervalMonthDayNano("86400000000000") |
+----------------------------------------+
| P1D                                    |
+----------------------------------------+
1 row in set. Query took 0.002 seconds.
```

it make sense to raise error while converting month unit, but for days should we fix it?

**Comment by tustvold** [comment]
Days aren't necessarily 24 hours, e.g. at a DST boundary

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/pull/4557 proposes changing the formatting of durations upstream, PTAL

File: datafusion/core/tests/sql/timestamp.rs (hunk 2 above)
Comment (alamb) [reply datafusion/core/tests/sql/timestamp.rs:1041]: I have filed two tickets of interest: 
* https://github.com/apache/arrow-datafusion/issues/7068 which tries to describes the change in semantics this PR does
* https://github.com/apache/arrow-datafusion/issues/7070 that proposes special casing the display of Duration, based on @waitingkuo 's feedback above

File: datafusion/physical-expr/src/format.rs [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow::util::display::{DurationFormat, FormatOptions};
+
+/// The default [`FormatOptions`] to use within DataFusion
+pub const DEFAULT_FORMAT_OPTIONS: FormatOptions<'static> =
```
Comment (tustvold) [inline datafusion/physical-expr/src/format.rs:21]: I wasn't really sure where to put this, I may move this into core so that it can be used by the assert_batches_eq macro :thinking:

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 10 above)
Comment (viirya) [reply datafusion/physical-expr/src/expressions/binary.rs:2433]: I think that it the rules for precision of decimal division are changed. I won't say previous rules are incorrect.

File: datafusion/core/tests/sqllogictests/test_files/decimal.slt (hunk 8 above)
Comment (viirya) [reply datafusion/core/tests/sqllogictests/test_files/decimal.slt:368]: I won't say this is not correct, this is more like a design decision to me. But this actually changes the precision of decimal arithmetic operations in DataFusion. For some, this might be a different change and not something could be implicitly accepted. I'd suggest there are some documents on these changes with new release including this.

Comment (tustvold) [reply datafusion/core/tests/sqllogictests/test_files/decimal.slt:368]: > But this actually changes the precision of decimal arithmetic operations in DataFusion

Correct, it changes the semantics of decimal division to match that of postgres, mysql and Hive. I've not done an exhaustive survey of other DBMS products, but the Hive semantics were decided based on a survey of the systems at the time. Other decimal operations are unaffected. This should be definitely communicated, especially if Spark does something different (possibly as a misreading of the Hive specification).

File: datafusion/physical-expr/src/format.rs (hunk 11 above)
Comment (alamb) [reply datafusion/physical-expr/src/format.rs:21]: How about `datafusion_common` as it seems like common functionality across moduels?

File: datafusion/core/tests/sqllogictests/test_files/dates.slt [hunk 12]
```diff
@@ -91,17 +91,19 @@ where d3_date > now() + '5 days';
...
 SELECT DATE '2023-04-09' - DATE '2023-04-02';
+----
+7 days 0 hours 0 mins 0 secs
 
 # DATE minus Timestamp
 query ?
 SELECT DATE '2023-04-09' - '2000-01-01T00:00:00'::timestamp;
 ----
-0 years 0 mons 8499 days 0 hours 0 mins 0.000000000 secs
+8499 days 0 hours 0 mins 0.000000000 secs
```
Comment (alamb) [inline datafusion/core/tests/sqllogictests/test_files/dates.slt:103]: much better

File: datafusion/core/tests/sqllogictests/test_files/decimal.slt (hunk 8 above)
Comment (alamb) [reply datafusion/core/tests/sqllogictests/test_files/decimal.slt:368]: I think spark is of the other major targets / comparison system -- I think @liukun4515  and maybe @yjshen  perhaps would have more context / understanding of how potential changes to the decimal division semantics may affect downstream users

If there is need for both semantics, maybe we could add a  "use_spark_decimal_division_rules" config option to DataFusion

**Review by alamb — COMMENTED** [review]
The duration display now looks good to me. Thank you @tustvold 

The only thing that is not 100% clear to me is if the decimal division changes have consensus. Does any one else have thoughts?

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
@viirya  / @waitingkuo  / @liukun4515  I wonder if you could verify this PR is acceptable (specifically the Decimal display differences as well as the change in decimal division semantics)

**Comment by Dandandan** [comment]
Does this also fix https://github.com/apache/arrow-datafusion/issues/6794 (see https://github.com/apache/arrow-datafusion/issues/5646#issuecomment-1663422469)

**Comment by tustvold** [comment]
It should do, but I haven't verified this

**Comment by alamb** [comment]
I have posted an announcement about this PR to the mailing list: https://lists.apache.org/thread/837ghhjh9gd1kvg0qqnmxqs0rm62x0xt

I will plan to merge it in Monday Aug 7 2023 unless I hear otherwise

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/pull/4640 proposes a further tweak to the decimal arithmetic rules, in light of #6828 I'm not sure this should be a blocker to getting this in, but defer to the consensus. Worst case we could do a patch release for arrow-rs, but I would rather avoid this if possible

**Comment by alamb** [comment]
I am going to update this PR and resolve conflicts

**Review by Dandandan — APPROVED** [review]

**Comment by alamb** [comment]
I am running this update against the IOx tests as one final check: https://github.com/influxdata/influxdb_iox/pull/8452

**Comment by alamb** [comment]
Our tests passed. 🚀  Onward

**Comment by alamb** [comment]
One final shoutout for this cleanup in this PR:

Screenshot 2023-08-08 at 12 56 17 PM

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/issues/4664 tracks further upstream work related to decimal arithmetic

**Comment by liukun4515** [comment]
> > @viirya / @waitingkuo / @liukun4515 I wonder if you could verify this PR is acceptable (specifically the Decimal display differences as well as the change in decimal division semantics)
> 
> I only concern the change of decimal division semantics could be an impact to some customers potentially if high precision decimal results are important to them. Ideally I'd like to get same precision results as Spark but under current design I think it might be tricky to have two different precision rules for decimal operations. It could possibly make DataFusion complicated on decimal operation too. So it may not be a good idea. Among existing systems, I also don't know if any provides alternative option regarding decimal precision rule.

@tustvold @alamb @viirya 
spark has the config to control the `precision loss` : https://github.com/apache/spark/blob/2be20e54a2222f6cdf64e8486d1910133b43665f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L246

**Comment by tustvold** [comment]
https://github.com/apache/arrow-rs/issues/4664#issuecomment-1676091528 summarises my thoughts on the issue having spent significant time on this. TLDR is we already support a higher precision than most commercial systems, and given this and the fact I personally have no use-cases for decimals, find it hard to justify spending more of my time on this. Ultimately if someone really wants true arbitrary precision decimal arithmetic a la postgres or spark, arrow is probably the wrong tool.

**Comment by viirya** [comment]
> @tustvold @alamb @viirya spark has the config to control the `precision loss` : https://github.com/apache/spark/blob/2be20e54a2222f6cdf64e8486d1910133b43665f/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/arithmetic.scala#L246

I know it. But so?
